pkg/planner/core/plan_stats.go

```diff
@@ -39,24 +40,83 @@ func (collectPredicateColumnsPoint) optimize(_ context.Context, plan LogicalPlan
 	predicateNeeded := variable.EnableColumnTracking.Load()
 	syncWait := plan.SCtx().GetSessionVars().StatsLoadSyncWait
 	histNeeded := syncWait > 0
-	predicateColumns, histNeededColumns := CollectColumnStatsUsage(plan, predicateNeeded, histNeeded)
+	predicateColumns, histNeededColumns, visitedPhysTblIDs := CollectColumnStatsUsage(plan, predicateNeeded, histNeeded)
 	if len(predicateColumns) > 0 {
 		plan.SCtx().UpdateColStatsUsage(predicateColumns)
 	}
-	if !histNeeded {
-		return plan, planChanged, nil
-	}
 
-	// Prepare the table metadata to avoid repeatedly fetching from the infoSchema below.
+	// Prepare the table metadata to avoid repeatedly fetching from the infoSchema below, and trigger extra sync/async
+	// stats loading for the determinate mode.
 	is := plan.SCtx().GetInfoSchema().(infoschema.InfoSchema)
+	statsHandle := domain.GetDomain(plan.SCtx()).StatsHandle()
 	tblID2Tbl := make(map[int64]table.Table)
+	physTblIDsWithNeededCols := intset.NewFastIntSet()
 	for _, neededCol := range histNeededColumns {
-		tbl, _ := infoschema.FindTableByTblOrPartID(is, neededCol.TableID)
+		physTblIDsWithNeededCols.Insert(int(neededCol.TableID))
+	}
+	visitedPhysTblIDs.ForEach(func(physicalTblID int) {
+		// 1. collect table metadata
+		tbl, _ := infoschema.FindTableByTblOrPartID(is, int64(physicalTblID))
 		if tbl == nil {
-			continue
+			return
 		}
-		tblID2Tbl[neededCol.TableID] = tbl
-	}
+		tblID2Tbl[int64(physicalTblID)] = tbl
+
+		// 2. handle extra sync/async stats loading for the determinate mode
+
+		// If we visited a table without getting any columns need stats (likely because there are no pushed down
+		// predicates), and we are in the determinate mode, we need to make sure we are able to get the "analyze row
+		// count" in getStatsTable(), which means any column/index stats are available.
+		if plan.SCtx().GetSessionVars().GetOptObjective() != variable.OptObjectiveDeterminate ||
+			// If we already collected some columns that need trigger sync laoding on this table, we don't need to
+			// additionally do anything for determinate mode.
+			physTblIDsWithNeededCols.Has(physicalTblID) ||
+			statsHandle == nil {
+			return
+		}
+		tblStats := statsHandle.GetTableStats(tbl.Meta())
+		if tblStats == nil || tblStats.Pseudo {
+			return
+		}
+		var colToTriggerLoad *model.TableItemID
+		for _, col := range tbl.Cols() {
+			if col.State != model.StatePublic || (col.IsGenerated() && !col.GeneratedStored) {
+				continue
+			}
+			if colStats := tblStats.Columns[col.ID]; colStats != nil {
+				// Choose the first column we meet to trigger stats loading.
+				if colToTriggerLoad == nil && colStats.IsStatsInitialized() {
+					colToTriggerLoad = &model.TableItemID{TableID: int64(physicalTblID), ID: col.ID, IsIndex: false}
+				}
+				// If any stats are already full loaded, we don't need to trigger stats loading on this table.
+				if colStats.IsFullLoad() {
+					colToTriggerLoad = nil
+					break
+				}
```

Comment:
Do we need to check it here since we have the check `physTblIDsWithNeededCols.Has(physicalTblID)`?

Reply:
They are different.
`physTblIDsWithNeededCols.Has(physicalTblID)` means we will trigger sync loading on some columns for this table now, no matter if there are already some usable stats.
`colStats.IsFullLoad()` means we already have some column stats that can be used and also implies that these stats are not needed by the predicates.